packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -712,33 +712,57 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     // case where the user moves the cursor to the end or beginning of the text
     // and then back up or down.
     if (downArrow || upArrow) {
-      // The caret offset gives a location in the upper left hand corner of
-      // the caret so the middle of the line above is a half line above that
-      // point and the line below is 1.5 lines below that point.
-      final double preferredLineHeight = _textPainter.preferredLineHeight;
-      final double verticalOffset = upArrow ? -0.5 * preferredLineHeight : 1.5 * preferredLineHeight;
-
-      final Offset caretOffset = _textPainter.getOffsetForCaret(TextPosition(offset: newSelection.extentOffset), _caretPrototype);
-      final Offset caretOffsetTranslated = caretOffset.translate(0.0, verticalOffset);
-      final TextPosition position = _textPainter.getPositionForOffset(caretOffsetTranslated);
-
-      // To account for the possibility where the user vertically highlights
-      // all the way to the top or bottom of the text, we hold the previous
-      // cursor location. This allows us to restore to this position in the
-      // case that the user wants to unhighlight some text.
-      if (position.offset == newSelection.extentOffset) {
-        if (downArrow) {
-          newSelection = newSelection.copyWith(extentOffset: _plainText.length);
-        } else if (upArrow) {
-          newSelection = newSelection.copyWith(extentOffset: 0);
+      if (lineModifier) {
+        if (upArrow) {
+          // Extend the selection to the beginning of the field.
+          final int upperOffset = math.max(
```

Comment:
can `newSelection` be `(-1, -1)`?

Reply:
According to `TextSelection.extentOffset`'s documentationn:

>   /// The offset at which the selection terminates.
  ///
  /// When the user uses the arrow keys to adjust the selection, this is the
  /// value that changes. Similarly, if the current theme paints a caret on one
  /// side of the selection, this is the location at which to paint the caret.
  ///
  /// Might be larger than, smaller than, or equal to base.

In the new implementation `baseOffset` may also channge (when <kbd>Shift</kbd> is pressed and `newSelection.extentOffset > newSelection.baseOffset`). Is that intended?

Reply:
For your first question, I was unable to figure out a way to get newSelection to be -1,-1 here, but I'm not confident that it never could be.  I've pushed a commit to make sure that case is handled as well (it's effectively treated like 0,0).

For your second question, yes that's intended, as tested on Mac's TextEdit.  Let me know if this doesn't cover what you mean:

  1. In a multi-line TextField with a bunch of text in it on several lines...
  2. Select some text in the middle so that the extent (the side that moves when you move the arrow keys) is ahead (to the right) of the base.
  3. Press shift+cmd+down to extend the selection to the end of the field.  The extent will be at the end of the field and the base will be in the same place.
  4. Press shift+cmd+up to extend the selection up to the beginning of the field so that everything is now selected. The extent will now be at the beginning of the field and the base will be at the end.